File: Algorithms/Graph/NumberOfIslands.cpp
#include <vector>
#include <iostream>
using namespace std;

class Solution {
public:
    int numIslands(vector<vector<char> >& grid) {
        if (grid.empty() || grid.size() == 0 || grid[0].size() == 0) {
            return 0;
        }
        
        int count = 0;
        const int rows = grid.size();
        const int cols = grid[0].size();
        for (int i = 0; i < rows; ++i) {
            for (int j = 0; j < cols; ++j) {
                if (grid[i][j] == '1') {
                    count++;
                    dfs(grid, i, j, rows, cols);
                }
            }
        }
        return count;
    }
    
    void dfs(vector<vector<char> >& grid, int i, int j, int rows, int cols) {
        if (i < 0 || i >= rows || j < 0 || j >= cols || grid[i][j] == '0') {
            return;
        }
        grid[i][j] = '0';
        dfs(grid, i + 1, j, rows, cols);
        dfs(grid, i - 1, j, rows, cols);
        dfs(grid, i, j + 1, rows, cols);
        dfs(grid, i, j - 1, rows, cols);
    }
};


int main() {
    vector<vector<char> > grid = {
                                {'1', '1', '0', '0', '0'},
                                {'1', '1', '0', '0', '0'},
                                {'0', '0', '1', '0', '0'},
                                {'0', '0', '0', '1', '1'}
                                };

    cout << grid.size() << endl;
    cout << grid[0].size() << endl;
    Solution s;
    int ret = s.numIslands(grid);
    cout << "ret: " << ret << endl;
    return 0;
}

File: Algorithms/Graph/NumberOfIslands.py
"""
Given a 2d grid map of ''1''s (land) and ''0''s (water), count the number of islands. An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically. You may assume all four edges of the grid are all surrounded by water.

Example '1':

'1''1''1''1''0'
'1''1''0''1''0'
'1''1''0''0''0'
'0''0''0''0''0'

Answer: '1'

Example 2:

'1''1''0''0''0'
'1''1''0''0''0'
'0''0''1''0''0'
'0''0''0''1''1'

Answer: 3
"""

import unittest


class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        Time complexity: O(V + E), Space complexity:
        """
        def sink(i, j):
            if 0 <= i < len(grid) and 0 <= j < len(grid[i]) and grid[i][j] == '1':
                grid[i][j] = '0'
                map(sink, (i + 1, i - 1, i, i), (j, j, j + 1, j - 1))
                return 1
            return 0
        return sum(sink(i, j) for i in range(len(grid)) for j in range(len(grid[i])))


class SolutionTest(unittest.TestCase):
    def setUp(self):
        self.s = Solution()

    def test1(self):
        grid = [
                ['1', '1', '1', '1', '0'],
                ['1', '1', '0', '1', '0'],
                ['1', '1', '0', '0', '0'],
                ['0', '0', '0', '0', '0']]
        self.assertEqual(self.s.numIslands(grid), 1)

    def test2(self):
        grid = [
                ['1', '1', '0', '0', '0'],
                ['1', '1', '0', '0', '0'],
                ['0', '0', '1', '0', '0'],
                ['0', '0', '0', '1', '1']]
        self.assertEqual(self.s.numIslands(grid), 3)


if __name__ == '__main__':
    unittest.main()
    """
    s = Solution()
    grid = [
        ['1', '1', '0', '0', '0'],
        ['1', '1', '0', '0', '0'],
        ['0', '0', '1', '0', '0'],
        ['0', '0', '0', '1', '1']]
    ret = s.numIslands(grid)
    print "ret: ", ret
    """
